src/System.IO.FileSystem/src/System/IO/Win32FileStreamAsyncResult.cs

```diff
@@ -96,232 +53,147 @@ unsafe private sealed class FileStreamAsyncResult : IAsyncResult
                 var ioCallback = s_IOCallback; // cached static delegate; delay initialized due to it being SecurityCritical
                 if (ioCallback == null) s_IOCallback = ioCallback = new IOCompletionCallback(AsyncFSCallback);
                 _overlapped = overlapped.Pack(ioCallback, bytes);
-
                 Debug.Assert(_overlapped != null, "Did Overlapped.Pack or Overlapped.UnsafePack just return a null?");
-#endif
             }
 
-            internal static FileStreamAsyncResult CreateBufferedReadResult(int numBufferedBytes, AsyncCallback userCallback, Object userStateObject, bool isWrite)
+            public Task<int> Task
             {
-                FileStreamAsyncResult asyncResult = new FileStreamAsyncResult(numBufferedBytes, userCallback, userStateObject, isWrite);
-                asyncResult.CallUserCallback();
-                return asyncResult;
+                get { return _completionSource.Task; }
             }
 
-            // This creates a synchronous Async Result. We should consider making this a separate class and maybe merge it with 
-            // System.IO.Stream.SynchronousAsyncResult
-            private FileStreamAsyncResult(int numBufferedBytes, AsyncCallback userCallback, Object userStateObject, bool isWrite)
+            internal NativeOverlapped* Overlapped
             {
-                _userCallback = userCallback;
-                _userStateObject = userStateObject;
-                _isWrite = isWrite;
-                _numBufferedBytes = numBufferedBytes;
+                [SecurityCritical]get { return _overlapped; }
             }
 
-            public Object AsyncState
+            bool IAsyncResult.IsCompleted
             {
-                get { return _userStateObject; }
+                get { return _completionSource.Task.IsCompleted; }
             }
 
-            public bool IsCompleted
+            object IAsyncResult.AsyncState
             {
-                get { return _isComplete; }
+                get { return ((IAsyncResult)_completionSource.Task).AsyncState; }
```

Comment:
This is always null, right, since you're not passing any state to the TaskCompletionSource's ctor?  If someone's actually using this property and expecting something else, then there's a bug here; otherwise, assuming you still need to implement the property, you can just return null, or even throw an exception or assert or something to ensure it's not being used.